### PR TITLE
botan3: init at 3.1.1

### PR DESCRIPTION
**Pull request by thillux**
###### Description of changes

botan was released in version 3 earlier this year. Version 3 most importantly adds TLS 1.3 and support for PQC algorithms. Reuse existing generic botan file. Drop patching of C++ standard, which was added for botan 2.0.1 some time ago and is no longer needed.

See: https://botan.randombit.net/news.html

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thillux** [comment]
Result of `nixpkgs-review pr 244212` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>biboumi</li>
    <li>botan2</li>
    <li>botan2.dev</li>
    <li>botan3</li>
    <li>botan3.dev</li>
    <li>chickenPackages_5.chickenEggs.botan</li>
    <li>corectrl</li>
    <li>ikiwiki-full</li>
    <li>keepassxc</li>
    <li>monotone</li>
    <li>qownnotes</li>
    <li>rnp</li>
    <li>rnp.dev</li>
    <li>rnp.lib</li>
    <li>softhsm</li>
    <li>step-kms-plugin</li>
  </ul>
</details>

File: pkgs/development/libraries/botan/3.0.nix [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{ callPackage, fetchpatch, ... } @ args:
+
+callPackage ./generic.nix (args // {
+  baseVersion = "3.1";
+  revision = "1";
+  sha256 = "sha256-MMhP6RmTapj+9TMfJGxiqiwOTSCFstRREgf2ogr6Oms=";
```
Comment (nikstur) [inline pkgs/development/libraries/botan/3.0.nix:6]:
```suggestion
  meta.badPlatforms = [ lib.platforms.darwin ];
```

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -20292,6 +20292,10 @@ with pkgs;
     inherit (darwin.apple_sdk.frameworks) CoreServices Security;
   };
 
+  botan3 = callPackage ../development/libraries/botan/3.0.nix {
+    inherit (darwin.apple_sdk.frameworks) CoreServices Security;
+  };
```
Comment (nikstur) [inline pkgs/top-level/all-packages.nix:20297]:
```suggestion
  botan3 = callPackage ../development/libraries/botan/3.0.nix {  };
```

**Review by nikstur — CHANGES_REQUESTED** [review]
Looks like botan3 doesn't build on darwin (macos) anymore: https://logs.ofborg.org/?key=nixos/nixpkgs.244212&attempt_id=20961051-b956-4926-975b-1ae6ce2a694c

I don't think that it fundamentally doesn't so lets add darwin to `meta.badPlatforms`. Unless you have a Mac and want to try to fix it of course.

**Comment by thillux** [comment]
> Looks like botan3 doesn't build on darwin (macos) anymore: https://logs.ofborg.org/?key=nixos/nixpkgs.244212&attempt_id=20961051-b956-4926-975b-1ae6ce2a694c
> 
> I don't think that it fundamentally doesn't so lets add darwin to `meta.badPlatforms`. Unless you have a Mac and want to try to fix it of course.

Botan 3 raised the bar on its dependencies quite a bit. It seems, that the default Clang 11.0 on MacOS is just too old for C++20. As long as botan2 is used by most tools, I would mark MacOS as bad platform and add a comment to reconsider this after MacOS gets a more recent compiler.

**Review by nikstur — APPROVED** [review]
LGTM, conditional on ofborg succeeding.

**Comment by 7c6f434c** [comment]
@ofborg eval

File: pkgs/development/libraries/botan/3.0.nix [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{ callPackage, fetchpatch, lib, ... } @ args:
+
+callPackage ./generic.nix (args // {
+  baseVersion = "3.1";
+  revision = "1";
+  sha256 = "sha256-MMhP6RmTapj+9TMfJGxiqiwOTSCFstRREgf2ogr6Oms=";
+  # reconsider removing this platform marking, when MacOS uses Clang 14.0+ by default.
+  meta.badPlatforms = [ lib.platforms.darwin ];
```
Comment (nikstur) [inline pkgs/development/libraries/botan/3.0.nix:8]: Ahh you would need to `overrrideAttrs` this derivation and add meta.badPlatforms there. My bad. Or you could add a `batPlatforms` flag to generic.nix

Comment (thillux) [reply pkgs/development/libraries/botan/3.0.nix]: I added a badPlatforms Flag.

**Pull request merged**